api/v1alpha1/ext_auth_types.go

```diff
@@ -0,0 +1,108 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import (
+	gwapiv1a2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+// ExtAuthServiceType specifies the types of External Authorization.
+// +kubebuilder:validation:Enum=GRPC;HTTP
+type ExtAuthServiceType string
+
+const (
+	// GRPC external authorization service.
+	GRPCExtAuthServiceType ExtAuthServiceType = "GRPC"
+
+	// HTTP external authorization service.
+	HTTPExtAuthServiceType ExtAuthServiceType = "HTTP"
+)
+
+// +kubebuilder:validation:XValidation:message="http must be specified if type is HTTP",rule="self.type == 'HTTP' ? has(self.http) : true"
+// +kubebuilder:validation:XValidation:message="grpc must be specified if type is GRPC",rule="self.type == 'GRPC' ? has(self.grpc) : true"
+// +kubebuilder:validation:XValidation:message="only one of grpc or http can be specified",rule="!(has(self.grpc) && has(self.http))"
+//
+// ExtAuth defines the configuration for External Authorization.
+type ExtAuth struct {
+	// Type decides the type of External Authorization.
+	// Valid ExtAuthServiceType values are "GRPC" or "HTTP".
+	// +kubebuilder:validation:Enum=GRPC;HTTP
+	// +unionDiscriminator
+	Type ExtAuthServiceType `json:"type"`
+
+	// GRPC defines the gRPC External Authorization service
+	// Only one of GRPCService or HTTPService may be specified.
+	GRPC *GRPCExtAuthService `json:"grpc,omitempty"`
+
+	// HTTP defines the HTTP External Authorization service
+	// Only one of GRPCService or HTTPService may be specified.
+	HTTP *HTTPExtAuthService `json:"http,omitempty"`
+
+	// AllowedHeaders defines the client request headers that will be included
+	// in the request to the external authorization service.
+	// Note: If not specified, the default behavior of different external authorization
+	// services is different. All headers will be included in the check request
+	// to a gRPC authorization server, whereas no headers will be included in the
+	// check request to an HTTP authorization server.
+	// +optional
+	AllowedHeaders []string `json:"allowedHeaders,omitempty"`
```

Comment:
thoughts on `AllowedExtAuthHeaders` ?

Reply:
Not sure about this: `ExtAuth` here could have multiple meaning

* Headers send to ExtAuth
* Headers from ExtAuth Response

The second one is wrong.

The most unambiguous name should be `AllowedClientHeadersToExtAuth`, but I guess `AllowedClientHeaders` should be clear enough?

Reply:
`AllowedClientHeaders` (from client) conflicts with `AllowedBackendHeaders` (to backend) :)

Reply:
`ClientHeadersToExtAuth` and `ExtAuthHeadersToBackend` ?

Reply:
Emissary's names for these ended up, in practice, often confusing people. For all that `HeadersToExtAuth` and `HeadersToBackend` may sound a bit excessive, I like that they will (hopefully) be very clear -- let's use them. 🙂